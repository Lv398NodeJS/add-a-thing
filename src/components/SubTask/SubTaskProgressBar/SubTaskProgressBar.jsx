import React, { Component } from 'react';
import { Row, Col, ProgressBar } from 'react-bootstrap';
import { getProgressBarVariant, getCompletedSubTasksPercent } from './subTaskProgressBarUtils';

class SubTaskProgressBar extends Component {
  render() {
    const { subtaskList = [] } = this.props;
    const completedSubTasksPercent = getCompletedSubTasksPercent(subtaskList);
    return (
      <Row className="my-2 mt-0 mx-0">
        <Col className="px-0 text-center">
          <span>{`Completed: ${completedSubTasksPercent}%`}</span>
          <ProgressBar
            className="subtask-progressbar"
            animated
            now={completedSubTasksPercent}
            variant={getProgressBarVariant(completedSubTasksPercent)}
          />
        </Col>
      </Row>
    );
  }
}

export default SubTaskProgressBar;
