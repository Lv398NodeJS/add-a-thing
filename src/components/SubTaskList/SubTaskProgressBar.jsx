import React, { Component } from 'react';
import { Row, Col, ProgressBar } from 'react-bootstrap';

class SubTaskAdd extends Component {
  render() {
    const { subtaskList } = this.props;
    const completedSubTasks = subtaskList.filter(subTask => subTask.completed);
    const completedSubTasksPercent = (completedSubTasks.length / subtaskList.length) * 100;
    const progressBarVariant = (percent) => {
      switch (true) {
        case percent === 100:
          return 'success';
        case percent > 50:
          return 'info';
        case percent > 30:
          return 'warning';
        default:
          return 'secondary';
      }
    };
    return (
      <Row className="justify-content-sm-center">
        <Col>
          <ProgressBar
            className="mb-3"
            animated
            now={completedSubTasksPercent}
            variant={progressBarVariant(completedSubTasksPercent)}
          />
        </Col>
      </Row>
    );
  }
}

export default SubTaskAdd;
