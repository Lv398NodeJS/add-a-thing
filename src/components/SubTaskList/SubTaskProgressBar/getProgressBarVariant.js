export default (percent) => {
  switch (true) {
    case percent === 100:
      return 'success';
    case percent > 50:
      return 'info';
    case percent > 30:
      return 'warning';
    default:
      return 'secondary';
  }
};
