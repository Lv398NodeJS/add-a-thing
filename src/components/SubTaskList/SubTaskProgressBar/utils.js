export const getProgressBarVariant = (percent = 0) => {
  switch (true) {
    case percent === 100:
      return 'success';
    case percent > 50:
      return 'info';
    case percent > 30:
      return 'warning';
    default:
      return 'secondary';
  }
};

export const getCompletedSubTasksPercent = (subtaskList = {}) => {
  if (!Object.keys(subtaskList).length) return 0;
  const completedSubTasks = subtaskList.filter(subTask => subTask.completed);
  const completedSubTasksPercent = (completedSubTasks.length / subtaskList.length) * 100;
  return Math.ceil(completedSubTasksPercent);
};
