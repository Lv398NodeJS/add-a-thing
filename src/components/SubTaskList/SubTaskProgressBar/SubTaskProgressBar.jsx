import React, { Component } from 'react';
import { Row, Col, ProgressBar } from 'react-bootstrap';
import getProgressBarVariant from './getProgressBarVariant';

class SubTaskProgressBar extends Component {
  render() {
    const { subtaskList = [] } = this.props;
    const completedSubTasks = subtaskList.filter(subTask => subTask.completed);
    const completedSubTasksPercent = (completedSubTasks.length / subtaskList.length) * 100;
    return (
      <Row className="justify-content-sm-center">
        <Col>
          <ProgressBar
            className="subtask-progressbar mb-3"
            animated
            now={completedSubTasksPercent}
            variant={getProgressBarVariant(completedSubTasksPercent)}
          />
        </Col>
      </Row>
    );
  }
}

export default SubTaskProgressBar;
