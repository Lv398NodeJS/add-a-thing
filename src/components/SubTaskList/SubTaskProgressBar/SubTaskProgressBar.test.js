import React from 'react';
import { shallow } from 'enzyme';
import SubTaskProgressBar from './SubTaskProgressBar';

describe('SubTaskProgressBar component', () => {
  it('should render correctly with no props', () => {
    const subTaskProgressBarComponent = shallow(<SubTaskProgressBar />);
    expect(subTaskProgressBarComponent).toMatchSnapshot();
  });

  it('should render progress bar "now" value properly', () => {
    const subtaskList = [{ completed: true }, { completed: false }];
    const subTaskProgressBarComponent = shallow(<SubTaskProgressBar subtaskList={subtaskList} />);
    expect(subTaskProgressBarComponent.find('.subtask-progressbar').props().now).toBe(50);
  });

  it('should render progressbar variant for 100% completed subtasks properly', () => {
    const subtaskList = [{ completed: true }];
    const subTaskProgressBarComponent = shallow(<SubTaskProgressBar subtaskList={subtaskList} />);
    expect(subTaskProgressBarComponent.find('.subtask-progressbar').props().variant).toBe(
      'success',
    );
  });

  it('should render progressbar variant for >50% completed subtasks properly', () => {
    const subtaskList = [{ completed: true }, { completed: true }, { completed: false }];
    const subTaskProgressBarComponent = shallow(<SubTaskProgressBar subtaskList={subtaskList} />);
    expect(subTaskProgressBarComponent.find('.subtask-progressbar').props().variant).toBe('info');
  });

  it('should render progressbar variant for < 30% completed subtasks properly', () => {
    const subtaskList = [{ completed: false }];
    const subTaskProgressBarComponent = shallow(<SubTaskProgressBar subtaskList={subtaskList} />);
    expect(subTaskProgressBarComponent.find('.subtask-progressbar').props().variant).toBe(
      'secondary',
    );
  });

  it('should render progressbar variant properly even if there is no subtasks in subtaskList', () => {
    const subtaskList = [];
    const subTaskProgressBarComponent = shallow(<SubTaskProgressBar subtaskList={subtaskList} />);
    expect(subTaskProgressBarComponent.find('.subtask-progressbar').props().variant).toBe(
      'secondary',
    );
  });
});
