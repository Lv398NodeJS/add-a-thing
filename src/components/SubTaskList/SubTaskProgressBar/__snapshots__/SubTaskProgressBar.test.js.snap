// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubTaskProgressBar component should render correctly and match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SubTaskProgressBar />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <ForwardRef
        as="div"
      >
        <ForwardRef(render)
          animated={true}
          className="subtask-progressbar mb-3"
          now={NaN}
          variant="secondary"
        />
      </ForwardRef>,
      "className": "justify-content-sm-center",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "as": "div",
        "children": <ForwardRef(render)
          animated={true}
          className="subtask-progressbar mb-3"
          now={NaN}
          variant="secondary"
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "animated": true,
          "className": "subtask-progressbar mb-3",
          "now": NaN,
          "variant": "secondary",
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": undefined,
          "displayName": "Bootstrap(ProgressBar)",
          "propTypes": undefined,
          "render": [Function],
        },
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "as": "div",
        },
        "displayName": "Col",
        "render": [Function],
      },
    },
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": undefined,
      "displayName": "Bootstrap(Row)",
      "propTypes": undefined,
      "render": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <ForwardRef
          as="div"
        >
          <ForwardRef(render)
            animated={true}
            className="subtask-progressbar mb-3"
            now={NaN}
            variant="secondary"
          />
        </ForwardRef>,
        "className": "justify-content-sm-center",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "div",
          "children": <ForwardRef(render)
            animated={true}
            className="subtask-progressbar mb-3"
            now={NaN}
            variant="secondary"
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "animated": true,
            "className": "subtask-progressbar mb-3",
            "now": NaN,
            "variant": "secondary",
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": undefined,
            "displayName": "Bootstrap(ProgressBar)",
            "propTypes": undefined,
            "render": [Function],
          },
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "as": "div",
          },
          "displayName": "Col",
          "render": [Function],
        },
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": undefined,
        "displayName": "Bootstrap(Row)",
        "propTypes": undefined,
        "render": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
